### PR TITLE
Discount changes

### PR DESCRIPTION
**Pull request by amityadav0**
## Description

-  Remove broker to get discount in route_exact_amount_in and route_exact_amount_out
- Change discount values in tier
- Add discount in perpetual
- Add tier to as dependency to amm and perpetual 

### What has Changed?

What specific problem were you aiming to address, and how did you successfully resolve it? If tests were not uploaded for this pull request or if coverage decreased, please provide an explanation for the change.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/elys-network/elys/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#testing)
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### Deployment Notes

Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.

### Screenshots and Videos

Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.

### CODE REVIEW

**Review by Wgil — COMMENTED** [review]
- Can you please share get estimation by denom queries using platinum tier for Prashant to have a look?
- Please look at the `broker_address` field on parameter module, we have to get rid of it as well.
- While doing the above, you will notice that broker address is also being used on perpetuals. Lets try to remove this logic as well which I think it's also related to discounts.

File: proto/elys/tier/userdata.proto [hunk 1]
```diff
@@ -18,4 +18,16 @@ message Pool {
...
     (gogoproto.nullable) = false
   ];
+}
+
+message MembershipTier {
+  string minimum_portfolio = 1 [
+    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
+    (gogoproto.nullable) = false
+  ];
+  string membership = 2;
```
Comment (avkr003) [inline proto/elys/tier/userdata.proto:28]: Let's make this ENUM type if we are defining in proto

File: proto/elys/tier/userdata.proto [hunk 2]
```diff
@@ -18,4 +18,16 @@ message Pool {
...
+  string discount = 3 [
+    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
```
Comment (avkr003) [inline proto/elys/tier/userdata.proto:30]: if you can add (cosmos_proto.scalar) = "cosmos.Dec", above this line

File: proto/elys/tier/userdata.proto [hunk 3]
```diff
@@ -18,4 +18,16 @@ message Pool {
     (gogoproto.customtype) = "cosmossdk.io/math.Int",
     (gogoproto.nullable) = false
   ];
+}
+
+message MembershipTier {
+  string minimum_portfolio = 1 [
+    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
```
Comment (avkr003) [inline proto/elys/tier/userdata.proto:25]: if you can add (cosmos_proto.scalar) = "cosmos.Dec", above this line.
You might need to import  "cosmos_proto/cosmos.proto";

**Comment by amityadav0** [comment]
@Wgil 
1. Shared on slack
2. I saw it, this is also integrated in perpetual open, close messages, need to check and remove
3. will remove it after this PR

Added following things
 Remove broker to get discount in route_exact_amount_in and route_exact_amount_out
Change discount values in tier
Add discount in perpetual
Add tier to as dependency to amm and perpetual

File: x/perpetual/keeper/query_open_estimation.go [hunk 4]
```diff
@@ -105,7 +106,7 @@ func (k Keeper) HandleOpenEstimation(ctx sdk.Context, req *types.QueryOpenEstima
 		//getting custody
 		if mtp.CollateralAsset == baseCurrency {
 			leveragedAmtTokenIn := sdk.NewCoin(mtp.CollateralAsset, leveragedAmount)
-			custodyAmount, slippage, weightBreakingFee, err = k.EstimateSwapGivenIn(ctx, leveragedAmtTokenIn, mtp.CustodyAsset, ammPool)
+			custodyAmount, slippage, weightBreakingFee, err = k.EstimateSwapGivenIn(ctx, leveragedAmtTokenIn, mtp.CustodyAsset, ammPool, mtp.Address)
```
Comment (avkr003) [inline x/perpetual/keeper/query_open_estimation.go:109]: The mtp address is empty here as it's just a query. Either we get address or we send default. If we get address from query, then in response need to send final fees

File: x/perpetual/keeper/query_open_estimation.go [hunk 5]
```diff
@@ -118,7 +119,7 @@ func (k Keeper) HandleOpenEstimation(ctx sdk.Context, req *types.QueryOpenEstima
 		//getting Liabilities
 		if mtp.CollateralAsset != baseCurrency {
 			amountIn := req.Collateral.Amount.ToLegacyDec().Mul(eta).TruncateInt()
-			liabilities, slippage, weightBreakingFee, err = k.EstimateSwapGivenOut(ctx, sdk.NewCoin(req.Collateral.Denom, amountIn), baseCurrency, ammPool)
+			liabilities, slippage, weightBreakingFee, err = k.EstimateSwapGivenOut(ctx, sdk.NewCoin(req.Collateral.Denom, amountIn), baseCurrency, ammPool, mtp.Address)
```
Comment (avkr003) [inline x/perpetual/keeper/query_open_estimation.go:122]: same as above

File: x/perpetual/keeper/query_open_estimation.go [hunk 6]
```diff
@@ -134,7 +135,7 @@ func (k Keeper) HandleOpenEstimation(ctx sdk.Context, req *types.QueryOpenEstima
 		// Collateral will be in base currency
 		amountOut := req.Collateral.Amount.ToLegacyDec().Mul(eta).TruncateInt()
 		tokenOut := sdk.NewCoin(baseCurrency, amountOut)
-		liabilities, slippage, weightBreakingFee, err = k.EstimateSwapGivenOut(ctx, tokenOut, mtp.LiabilitiesAsset, ammPool)
+		liabilities, slippage, weightBreakingFee, err = k.EstimateSwapGivenOut(ctx, tokenOut, mtp.LiabilitiesAsset, ammPool, mtp.Address)
```
Comment (avkr003) [inline x/perpetual/keeper/query_open_estimation.go:138]: same as above

File: x/tier/keeper/portfolio.go [hunk 7]
```diff
@@ -24,6 +24,29 @@ import (
...
+		Discount:         sdkmath.LegacyMustNewDecFromStr("0.1"),
+		Membership:       "silver",
+		MinimumPortfolio: sdkmath.LegacyMustNewDecFromStr("50000"),
+	}
+	gold = types.MembershipTier{
+		Discount:         sdkmath.LegacyMustNewDecFromStr("0.2"),
+		Membership:       "gold",
+		MinimumPortfolio: sdkmath.LegacyMustNewDecFromStr("250000"),
+	}
+	platinum = types.MembershipTier{
```
Comment (avkr003) [inline x/tier/keeper/portfolio.go:43]: Lets make it exportable, might be useful in other modules and keep it in tier/types package

File: x/tradeshield/keeper/pending_spot_order_test.go [hunk 8]
```diff
@@ -10,12 +10,21 @@ import (
...
 	ammtypes "github.com/elys-network/elys/x/amm/types"
+	tiertypes "github.com/elys-network/elys/x/tier/types"
 	"github.com/elys-network/elys/x/tradeshield/keeper"
 	"github.com/elys-network/elys/x/tradeshield/types"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
+var (
+	gold = tiertypes.MembershipTier{
```
Comment (avkr003) [inline x/tradeshield/keeper/pending_spot_order_test.go:21]: Move it to tier / types

**Review by avkr003 — CHANGES_REQUESTED** [review]

File: x/perpetual/keeper/query_open_estimation.go (hunk 4 above)
Comment (amityadav0) [reply x/perpetual/keeper/query_open_estimation.go:109]: yes, I have handled it later. we use default value from tier module

**Review by avkr003 — APPROVED** [review]

File: x/perpetual/keeper/query_open_estimation.go (hunk 4 above)
Comment (Wgil) [reply x/perpetual/keeper/query_open_estimation.go:109]: quick question, should this be here `mtp.Address` or user address?

Comment (avkr003) [reply x/perpetual/keeper/query_open_estimation.go:109]: @amityadav0 @Wgil user address

Comment (amityadav0) [reply x/perpetual/keeper/query_open_estimation.go:109]: mtp.address is user address when opening a position, mtp.address is set to msg.creator

**Pull request merged**